lib/mayaUsd/ufe/UsdUndoRenameCommand.cpp

```diff
@@ -84,105 +105,83 @@ UsdSceneItem::Ptr UsdUndoRenameCommand::renamedItem() const
     auto defaultPrimPath = _stage->GetDefaultPrim().GetPath();
 
-    // handle unique name for _newName
-    TfToken::HashSet childrenNames;
-    for (auto child : prim.GetParent().GetChildren()){
-        childrenNames.insert(child.GetName());
-    }
-    if (childrenNames.find(TfToken(_newName)) != childrenNames.end()){
-        _newName = uniqueName(childrenNames, _newName);
-    }
+    // 1- open a changeblock to delay sending notifications.
+    // 2- update the Internal References paths (if any) first
+    // 3- set the new name 
+    {
+        SdfChangeBlock changeBlock;
 
-    // names are not allowed to start to digit numbers
-    if(std::isdigit(_newName.at(0))){
-        _newName = prim.GetName();
-    }
+        const UsdPrim& prim = _stage->GetPrimAtPath(_ufeSrcItem->prim().GetPath());
 
-    // all special characters are replaced with `_`
-    const std::string specialChars{"~!@#$%^&*()-=+,.?`':{}|<>[]/ "};
-    std::replace_if(_newName.begin(), _newName.end(), [&](auto c){
-        return std::string::npos != specialChars.find(c);
-    }, '_');
+        auto newSceneItem = createSiblingSceneItem(_ufeSrcItem->path(), _newName);
+        auto newUfePath = Ufe::Path(newSceneItem->path().getSegments()[1]);
+        bool status = MayaUsdUtils::updateInternalReferencesPath(prim, SdfPath(newUfePath.string()));
+        if (!status) {
+            return false;
+        }
 
-    // set prim's name
-    // XXX: SetName successfuly returns true but when you examine the _prim.GetName()
-    // after the rename, the prim name shows the original name HS, 6-May-2020.
-    bool status = primSpec->SetName(_newName);
-    if (!status) {
-        return false;
+        // set the new name
+        auto primSpec = MayaUsdUtils::getPrimSpecAtEditTarget(prim);
+        status = primSpec->SetName(_newName);
+        if (!status) {
+            return false;
+        }
     }
 
     // the renamed scene item is a "sibling" of its original name.
     _ufeDstItem = createSiblingSceneItem(_ufeSrcItem->path(), _newName);
 
-    sendNotification<Ufe::ObjectRename>(_ufeDstItem, _ufeSrcItem->path());
-
-    // SdfLayer is a "simple" container, and all it knows about defaultPrim is that it is a piece of token-valued layer metadata.  
-    // It is only the higher-level Usd and Pcp modules that know that it is identifying a prim on the stage.  
-    // One must use the SdfLayer API for setting the defaultPrim when you rename the prim it identifies.
-    if(primPath == defaultPrimPath){
+    // update stage's default prim
+    if(_ufeSrcItem->prim().GetPath() == defaultPrimPath) {
         _stage->SetDefaultPrim(_ufeDstItem->prim());
     }
 
-    // update internal references
-    status = MayaUsdUtils::updateInternalReferences(prim, _ufeDstItem->prim());
-    if (!status) {
-        return false;
-    }
+    // send notification to update UFE data model
+    sendNotification<Ufe::ObjectRename>(_ufeDstItem, _ufeSrcItem->path());
 
     return true;
```

Comment:
- Updating the internal references paths should happen first before the primSpec->SetName.

- Tidy up renameRedo a little bit and moved the unique name logics to the constructor.